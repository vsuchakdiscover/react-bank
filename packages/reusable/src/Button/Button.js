import React from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import styles from "./Button.module.scss";

export const BUTTON_TYPES = {
  PRIMARY: "primary",
  GHOST: "ghost",
  LINK: "link",
  TOOLTIP: "tooltip",
  CLOSEX: "closex",
};

/** Reusable Button */
const Button = React.forwardRef(
  ({ buttonStyle, children, className, adobeEvent, ...props }, ref) => {
    if (adobeEvent && props.type === "submit")
      throw new Error(
        "For safety, Button doesn't support tracking form submit button clicks. Why? Because forms can also be submitted via the enter key. So instead, remove the adobeEvent prop from this component and track the event using the form's submit handler."
      );
    function handleClick(e) {
      if (props.href) window.location.href = props.href;
      if (props.onClick) props.onClick(e);
    }

    return (
      <button
        // This attribute is read by LogClicks which wraps the entire app as a global click handler that runs *before* any child onClick.
        data-track={adobeEvent}
        className={cx(styles.root, styles[buttonStyle], className)}
        onClick={handleClick}
        ref={ref}
        {...props}
      >
        {children}
      </button>
    );
  }
);

Button.displayName = "Button";

Button.propTypes = {
  /** Event code sent to Adobe for tracking purposes on click.
   * For safety, this is NOT supported for type="submit" because form submissions
   * should be tracked in the form submit handler since the user may not click the
   * submit button to submit the form. */
  adobeEvent: PropTypes.string,

  /** Attribute that identifies the element(by id) that the button activates */
  "aria-controls": PropTypes.string,

  /** Describes state of element toggled */
  "aria-expanded": PropTypes.oneOf(["true", "false"]),

  /** Accessiblity attribute for buttons without semantic context */
  "aria-label": PropTypes.string,

  /** Button style */
  buttonStyle: PropTypes.oneOf([
    BUTTON_TYPES.PRIMARY,
    BUTTON_TYPES.CLOSEX,
    BUTTON_TYPES.GHOST,
    BUTTON_TYPES.LINK,
    BUTTON_TYPES.TOOLTIP,
  ]).isRequired,

  /** Child elements to render inside button */
  children: PropTypes.any,

  /** Class to apply */
  className: PropTypes.string,

  /** Set to true to disable the button */
  disabled: PropTypes.bool,

  /** URL to load on click */
  href: PropTypes.string,

  /** Function to call on click */
  onClick: PropTypes.func,

  /** Function to call on key press */
  onKeyPress: PropTypes.func,

  /** Ref */
  ref: PropTypes.object,

  /** Button type */
  type: PropTypes.oneOf(["button", "submit"]).isRequired,
};

Button.defaultProps = {
  buttonStyle: BUTTON_TYPES.PRIMARY,
  disabled: false,
  type: "button",
};

export default Button;
