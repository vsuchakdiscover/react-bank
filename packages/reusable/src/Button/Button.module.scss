@import "src/scss/variables";

.root {
  display: inline-block;
  min-height: 36px;
  text-decoration: none;
  border-radius: 4px;
  min-width: 74px;
  padding: 4px 20px 0;
  border: none;
  font-family: "Meta Web Bold";

  @media #{$mobile_only} {
    width: 100%;
  }
}

.primary {
  background: $orange;
  color: $white;
  font-size: 19px;

  &:hover,
  &:focus {
    background: #e94d00;
  }
}

.ghost {
  background: $white;
  border: 2px solid $links;
  color: $links;

  &:hover,
  &:focus {
    background: $legacy-bg-gray;
  }
}

.link {
  background: transparent;
  border: none;
  color: $links;
  font-size: 15px;
  min-height: auto;
  padding: 0;
  font-family: "Meta Web Normal";

  &:hover,
  &:focus {
    background: transparent;
    text-decoration: underline;
  }

  @media #{$mobile_only} {
    width: auto;
  }
}

.closex {
  background: transparent;
  color: $links;
  border: none;
  min-height: auto;
  padding: 20px;
  font-family: "Arial";
  min-width: auto;
  position: relative;
  border-radius: 0;

  &:after {
    content: "";
    position: absolute;
    height: 15px;
    width: 1px;
    background: $links;
    display: block;
    top: 13px;
    left: 20px;
    transform: rotate(45deg);
  }

  &:before {
    content: "";
    position: absolute;
    width: 15px;
    height: 1px;
    background: $links;
    display: block;
    top: 20px;
    left: 13px;
    transform: rotate(45deg);
  }

  @media #{$mobile_only} {
    width: auto;
  }
}

.tooltip {
  display: flex;
  min-width: 20px;
  padding-left: 0;
  padding-right: 0;
  margin-left: 10px;
  color: $links;
  width: auto;
  background: none;

  &:hover,
  &:focus {
    background: transparent;
    text-decoration: underline;
  }
}
