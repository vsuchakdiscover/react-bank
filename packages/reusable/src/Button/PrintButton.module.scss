.btn {
  text-align: right;
  font-size: 15px;
}

.icon {
  display: inline-block;
  vertical-align: middle;
  margin-right: 8px;
}
