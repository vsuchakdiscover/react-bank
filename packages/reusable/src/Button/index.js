// Barrel for shortening imports
export { default, BUTTON_TYPES } from "./Button";
