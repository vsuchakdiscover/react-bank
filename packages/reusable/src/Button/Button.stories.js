import React from "react";
import Button, { BUTTON_TYPES } from "./Button";
import { text } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";

export default { title: "Buttons/Button" };

export const defaultExample = () => (
  <Button onClick={action("clicked")} onKeyPress={action("keypressed")}>
    {text("button text", "Your Average Button")}
  </Button>
);

export const href = () => (
  <Button href="http://www.discover.com">
    {text("button text", "Go to Discover.com")}
  </Button>
);

export const ghost = () => (
  <Button
    buttonStyle={BUTTON_TYPES.GHOST}
    onClick={action("clicked")}
    onKeyPress={action("keypressed")}
  >
    {text("button text", "It's a Ghost!")}
  </Button>
);

export const submit = () => (
  <Button
    onClick={action("clicked")}
    onKeyPress={action("keypressed")}
    type="submit"
  >
    {text("button text", "Submit This")}
  </Button>
);

export const link = () => (
  <Button
    buttonStyle={BUTTON_TYPES.LINK}
    onClick={action("clicked")}
    onKeyPress={action("keypressed")}
    type="button"
  >
    {text("button text", "Modal Link Style")}
    <span className="sr-only"> opens a modal window</span>
  </Button>
);

export const modalCloseX = () => (
  <Button
    aria-label="Close modal dialog"
    buttonStyle={BUTTON_TYPES.CLOSEX}
    onClick={action("clicked")}
    onKeyPress={action("keypressed")}
    type="button"
  />
);

export const tooltip = () => (
  <div className="col-12">
    <Button
      aria-label="Close modal dialog"
      buttonStyle={BUTTON_TYPES.TOOLTIP}
      onClick={action("clicked")}
      onKeyPress={action("keypressed")}
      type="button"
    >
      Tooltip Button
    </Button>
  </div>
);
