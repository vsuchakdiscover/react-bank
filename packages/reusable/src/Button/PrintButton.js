import React from "react";
import PropTypes from "prop-types";
import Button, { BUTTON_TYPES } from "./Button";
import PrintIcon from "../Icons/PrintIcon";
import styles from "./PrintButton.module.scss";
import ReactToPrint from "react-to-print";
import cx from "classnames";

/** Print button that optionally automatically opens a print window containing the content passed via ref 
    The forwardRef is necessary to support passing a reference to the element to print to react-to-print
    If ref isn't provided, it's just a plain print button.
*/
const PrintButton = React.forwardRef((props, ref) => {
  function getButton() {
    return (
      <Button
        buttonStyle={BUTTON_TYPES.LINK}
        className={cx(styles.button, props.className)}
        onClick={props.onClick}
        style={{ fontSize: props.fontSize + "px" }}
      >
        {!props.hideIcon && <PrintIcon className={styles.icon} />}
        {props.label}
      </Button>
    );
  }

  return ref ? (
    <ReactToPrint
      content={() => ref.current}
      onAfterPrint={props.onAfterPrint}
      onBeforeGetContent={props.onBeforeGetContent}
      onBeforePrint={props.onBeforePrint}
      onPrintError={props.onPrintError}
      trigger={() => getButton()}
    />
  ) : (
    getButton()
  );
});

PrintButton.propTypes = {
  /** Classname to apply to the Button */
  className: PropTypes.string,

  /** Shows/Hides the Print Icon */
  hideIcon: PropTypes.bool,

  /** Print Button label */
  label: PropTypes.string,

  /** Callback function that triggers after print */
  onAfterPrint: PropTypes.func,

  /** Callback function that triggers before the library gathers the page's content. Either returns void or a Promise. This can be used to change the content on the page before printing. */
  onBeforeGetContent: PropTypes.func,

  /** Callback function that triggers before print. Either returns void or a Promise. Note: this function is run immediately prior to printing, but after the page's content has been gathered. To modify content before printing, use onBeforeGetContent instead. */
  onBeforePrint: PropTypes.func,

  /** Function to call when the button is clicked */
  onClick: PropTypes.func,

  /** Callback function that will be called if there is a printing error serious enough that printing cannot continue. Currently limited to Promise rejections in onBeforeGetContent or onBeforePrint. Use this to attempt to print again. errorLocation will tell you in which callback the Promise was rejected. */
  onPrintError: PropTypes.func,

  /** Font size */
  fontSize: PropTypes.number,
};

PrintButton.defaultProps = {
  label: "Print",
  hideIcon: false,
  fontSize: 15,
};

PrintButton.displayName = "PrintButton";

export default PrintButton;
