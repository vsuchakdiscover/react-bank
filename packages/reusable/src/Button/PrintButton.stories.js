import React, { useRef } from "react";
import PrintButton from "./PrintButton";

export default { title: "Buttons/PrintButton" };

function RefExample() {
  const ref = useRef();

  return (
    <>
      <p>
        If you only want to print part of the page, you can pass a ref to the
        content to print:
      </p>
      <p ref={ref}>
        When the print button is clicked, only this paragraph will be displayed
        in the print window since the ref is pointed to this paragraph. Note
        that ref is optional. Without a ref, you must call window.print
        yourself.
      </p>
      <PrintButton ref={ref} />
    </>
  );
}

export const defaultExample = () => <PrintButton />;

export const withOnClick = () => (
  <PrintButton fontSize={16} onClick={() => alert("clicked")} />
);

export const withRefToItemToPrint = () => <RefExample />;
