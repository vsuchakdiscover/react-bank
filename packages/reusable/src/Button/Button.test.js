import React from "react";
import Button from "./Button";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer.create(<Button>Example text</Button>).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders as disabled when disabled prop is specified", () => {
  const tree = renderer.create(<Button disabled>Example text</Button>).toJSON();
  expect(tree).toMatchSnapshot();
});
