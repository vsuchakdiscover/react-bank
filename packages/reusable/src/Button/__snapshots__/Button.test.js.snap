// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as disabled when disabled prop is specified 1`] = `
<button
  className="root primary"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  Example text
</button>
`;

exports[`renders correctly 1`] = `
<button
  className="root primary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Example text
</button>
`;
